src/lyra/abstract_domains/string/character_domain.py

```diff
@@ -0,0 +1,263 @@
+
+    def __repr__(self):
+        if self.is_bottom():
+            return "⊥"
+
+        if self.is_top():
+            return "T"
+
+        return "(<{}>, <{}>)".format(self.certainly, self.maybe)
+
+    @copy_docstring(BottomMixin.is_top)
+    def is_top(self) -> bool:
+        return not self.is_bottom() and self.certainly == set() and self.maybe == _alphabet
+
+    @copy_docstring(BottomMixin._less_equal)
+    def _less_equal(self, other: 'CharacterLattice') -> bool:
+        return self.maybe.issuperset(other.maybe) and self.certainly.issubset(other.certainly)
+
+    @copy_docstring(BottomMixin._join)
+    def _join(self, other: 'CharacterLattice') -> 'CharacterLattice':
+        maybe = self.certainly.intersection(other.certainly)
+        certainly = self.maybe.intersection(other.maybe)
+        return self._replace(CharacterLattice(maybe, certainly))
+
+    @copy_docstring(BottomMixin._meet)
+    def _meet(self, other: 'CharacterLattice'):
+        maybe = self.certainly.union(other.certainly)
+        certainly = self.maybe.intersection(other.maybe)
+        return self._replace(CharacterLattice(maybe, certainly))
+
+    @copy_docstring(BottomMixin._widening)
+    def _widening(self, other: 'CharacterLattice'):
+        return self._join(other)
+
+    def concat(self, other: 'CharacterLattice') -> 'CharacterLattice':
+        """
+            Define semantics for the string concatenation operation.
+            Both the **certainly** and **maybe** sets are joined
+        :param other: lattice element to be concatenated with the current element
+        :return: the result of the concatenation
+        """
+        certainly = self.certainly.union(other.certainly)
+        maybe = self.certainly.union(other.certainly)
+        return self._replace(certainly, maybe)
+
+
+class CharacterState(Store, State):
+
+    """
+        An implementation of the Character Inclusion Domain.
+
+        Maps from every string variable to a tuple of two sets: one set of characters that
+        are **certainly** included in the string and a set of characters that are **maybe**
+        included in the string. The string cannot contain characters from outside the **maybe**
+        set and must contain all characters in **certainly**. Therefore,
+        the **certainly** set is a subset of the **maybe** set.
+
+    """
+
+    class ConditionEvaluator(ExpressionVisitor):
+
+        def visit_Literal(self, expr, state=None, eval=None):
+            if expr.typ == StringLyraType():
+                return CharacterLattice(set(expr.val), set(expr.val))
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_Input(self, expr, state=None, eval=None):
+            raise NotImplementedError("Condition evalutator not supported for {}".format(expr))
+
+        def visit_VariableIdentifier(self, expr, state=None, eval=None):
+            return state.store[expr]
```

Comment:
With this you risk side-effect modifications via aliasing. I would do something similar to the interval domain:

```
            if expr in evaluation:
                return evaluation  # nothing to be done
            evaluation[expr] = deepcopy(state.store[expr])
            return evaluation
```